swarm/storage/mru/lookup/lookup.go

```diff
@@ -0,0 +1,199 @@
+// lookup process entirely.
+type ReadFunc func(epoch Epoch, now uint64) (interface{}, error)
+
+// NoClue is a hint that can be provided when the Lookup caller does not have
+// a clue about where the last update may be
+var NoClue = Epoch{}
+
+func getBaseTime(t uint64, level uint8) uint64 {
+	return t & (maxuint64 << level)
+}
+
+// Hint creates a hint based only on the last known update time
+func Hint(last uint64) Epoch {
+	return Epoch{
+		Time:  last,
+		Level: defaultLevel,
+	}
+}
+
+func getNextLevel(last Epoch, now uint64) uint8 {
+	mix := (last.Base() ^ now) | (1 << (last.Level - 1))
+	if mix > (maxuint64 >> (64 - highestLevel - 1)) {
+		return highestLevel
+	}
+	mask := uint64(1 << (highestLevel))
+
+	for i := uint8(highestLevel); i > lowestLevel; i-- {
+		if mix&mask != 0 {
+			return i
+		}
+		mask = mask >> 1
+	}
+	return 0
+}
+
+// GetNextEpoch returns the epoch where the next update should be located
+// according to where the previous update was
+// and what time it is now.
+func GetNextEpoch(last Epoch, now uint64) Epoch {
+	if last == NoClue {
+		return GetFirstEpoch(now)
+	}
+	level := getNextLevel(last, now)
+	return Epoch{
+		Level: level,
+		Time:  now,
+	}
+}
+
+// GetFirstEpoch returns the epoch where the first update should be located
+// based on what time it is now.
+func GetFirstEpoch(now uint64) Epoch {
+	return Epoch{Level: highestLevel, Time: now}
+}
+
+// Lookup finds the update with the highest timestamp that is smaller or equal than 'now'
+// It takes a hint which should be the epoch where the last known update was
+// If you don't know in what epoch the last update happened, simply submit lookup.NoClue
+// read() will be called on each lookup attempt
+// Returns an error only if read() returns an error
+// Returns nil if an update was not found
+func Lookup(now uint64, hint Epoch, read ReadFunc) (value interface{}, err error) {
+	var lastFound interface{}
+	var epoch Epoch
+	if hint == NoClue {
+		hint = Epoch{Time: 0, Level: highestLevel}
+	}
+
+	t := now
+
+	for {
+		epoch = GetNextEpoch(hint, t)
+		value, err = read(epoch, now)
+		if err != nil {
+			return nil, err
+		}
+		if value != nil {
+			lastFound = value
+			if epoch.Level == lowestLevel || epoch.Equals(hint) {
+				return value, nil
+			}
+			hint = epoch
+		} else {
+			if epoch.Base() == hint.Base() {
+				if lastFound != nil {
+					return lastFound, nil
+				}
+				// we have reached the hint itself
+				// check it out
+				value, err = read(hint, now)
+				if err != nil {
+					return nil, err
+				}
+				if value != nil {
+					return value, nil
+				}
+				// bad hint.
+				epoch = hint
+				hint = Epoch{Time: 0, Level: highestLevel}
+			}
+			base := epoch.Base()
+			if base == 0 {
+				return nil, nil
+			}
+			t = base - 1
+		}
+	}
+}
+
+// Lookup2 is a slower alternative lookup algorithm
+func Lookup2(now uint64, hint Epoch, read ReadFunc) (value interface{}, err error) {
```

Comment:
My `grep`ing suggests that this is not in use. If that's the case, please consider removing it.

Reply:
Removed. I will miss this guy, it was the first approach. 

The better one (now named `FluzCapacitorAlgorithm` actually occurred to me while I was writing the PR description and trying to figure out how on earth explain the other one.

I have separated the search algorithm function signature so different algorithms can be plugged and the same testing/benchmark functions can be used for possible future, hopefully faster, algorithms. The challenge is on!

Reply:
I like that. Good work.